docs/development/zod.md

```diff
@@ -0,0 +1,291 @@
+### Use `.transform()` method to process parsed data
+
+Schema validation helps to be more confident with types during downstream data transformation.
+
+You can go even further and perform some transformations as the part of schema itself.
+
+This is an example of **undesired** data transformation:
+
+```ts
+const Box = z.object({
+  width: z.number(),
+  height: z.number(),
+  length: z.number(),
+});
+
+const { width, height, length } = Box.parse(input);
+const volume = width * height * length;
+```
+
+Instead, use more idiomatic `.tranform()` method:
+
+```ts
+const Volume = z
+  .object({
+    width: z.number(),
+    height: z.number(),
+    length: z.number(),
+  })
+  .transform(({ width, height, length }) => width * height * length);
+
+Volume.parse({
+  width: 10,
+  height: 20,
+  length: 125,
+});
+```
+
+### Separate validation from transformation
+
+When parsing third party data, we are typically doing the following:
+
+- Validating that the data is correct/sufficient to use
+- Transforming it into a standardized format for internal usage
+
+Although it's not a strict requirement, your code will be cleaner if you perform the validation step first and then follow with transformation.
+
+### Stick to permissive behavior when possible
+
+Zod schemas are strict, and even some insufficient field is incorrect, the whole data will be treated like malformed.
+This could lead to cases when Renovate could've continued processing, but didn't.
+
+Remember: our goal is not to validate that data corresponds to any official specifications, but rather to ensure that the data is enough for Renovate to use.
+
+There are some techniques to make it more permissive to the input data.
+
+#### Use `.catch()` to force default values
+
+```ts
+const Box = z.object({
+  width: z.number().catch(10),
+  height: z.number().catch(10),
+});
+
+const box = Box.parse({ width: 10, height: null });
+// => { width: 10, height: 10 }
+```
+
+#### Use `LooseArray` and `LooseRecord` to filter out incorrect values from collections
+
+Suppose you want to validate an array and retain only number values in it.
+By using only methods provided by `zod` library, you'll have to write something like this:
+
+```ts
+const OnlyNumbers = z
+  .array(z.union([z.number(), z.null()]).catch(null))
+  .transform((xs) => xs.filter((x): x is number => x !== null));
+```
+
+While the problem is common, the code is quite complicated.
+
+Instead, you should use `LooseArray` and `LooseRecord` helpers provided in `schema-utils.ts`:
+
+```ts
+const OnlyNumbers = LooseArray(z.number());
+
+const xs = OnlyNumbers.parse([1, 2, null, 3, 'foobar', 4]);
+// => [1, 2, 3, 4]
+```
+
+### Combining with `Result` class
+
+Class `Result` (and also `AsyncResult`) represents the result of an operation, e.g. `Result.ok(200)` or `Result.err(404)`.
+
+It supports `.transform()` method, which is similar to `zod`'s one.
+
+It also supports `.onResult()` and `.onError()` methods for side-effectful result inspection.
```

Comment:
"Side-effectful result inspection" means nothing to me. Are you sure the developers reading this know what this means? I would prefer some simpler words here. 😄

Reply:
I think we're ok to leave it in